tidy3d/components/simulation.py

```diff
@@ -3777,23 +3777,26 @@ def _check_bloch_vec(
         if bloch_vec != expected_bloch_vec:
             test_val = np.real(expected_bloch_vec - bloch_vec)
 
-            if np.isclose(test_val % 1, 0) and not np.isclose(test_val, 0):
-                # the given Bloch vector is offset by an integer
-                log.warning(
-                    f"The wave vector of source '{source.name}' along dimension "
-                    f"'{dim}' is equal to the Bloch vector of the simulation "
-                    "boundaries along that dimension plus an integer reciprocal "
-                    "lattice vector. If using a 'DiffractionMonitor', diffraction "
-                    "order 0 will not correspond to the angle of propagation "
-                    "of the source. Consider using 'BlochBoundary.from_source()'.",
-                    custom_loc=["boundary_spec", "xyz"[dim]],
-                )
-            elif not np.isclose(test_val % 1, 0):
+            test_val_is_int = np.isclose(test_val, np.round(test_val))
+            src_name = f" '{source.name}'" if source.name else ""
+
+            # if test_val_is_int and not np.isclose(test_val, 0):
+            #     # the given Bloch vector is offset by an integer
+            #     log.warning(
+            #         f"The wave vector of source{src_name} along dimension "
+            #         f"'{dim}' is equal to the Bloch vector of the simulation "
+            #         "boundaries along that dimension plus an integer reciprocal "
+            #         "lattice vector. If using a 'DiffractionMonitor', diffraction "
+            #         "order 0 will not correspond to the angle of propagation "
+            #         "of the source. Consider using 'BlochBoundary.from_source()'.",
+            #         custom_loc=["boundary_spec", "xyz"[dim]],
+            #     )
```

Comment:
The way I read this, this was specifically added to avoid confusion when using such a source + a DiffractionMonitor. How about we only warn in that case then, i.e. check if a DiffractionMonitor is present rather than say "if using a diffraction monitor". This should fix your adjoint-specific annoyance with these warnings.

I remember at least one user was confused about this at the time.

Reply:
So I tried doing this, but the way our validators are set up in simulation, these don't have access to the `monitors` unfortunately. If I try to change anything, it breaks a ton of tests.. not sure if it's better to try to just rearrange things in simulation or simply remove the warning

Reply:
I came up with a solution that works, just added a separate validator for "monitors" and configured things to only check this condition if DiffractionMonitors are present there. seems to work properly.